sound/soc/sof/intel/hda-dsp.c

```diff
@@ -338,49 +338,155 @@ static int hda_dsp_send_pm_gate_ipc(struct snd_sof_dev *sdev, u32 flags)
-	u8 value;
+	int ret = 0;
+
+	mutex_lock(&hda->ps_mutex);
+	/* return if not in D0I3 */
+	if (hda->dsp_power_state != SOF_DSP_STATE_D0I3) {
+		dev_vdbg(bus->dev, "%s dsp power state already in %d\n",
+			 __func__, hda->dsp_power_state);
+		goto exit;
+	}
 
 	/* Write to D0I3C after Command-In-Progress bit is cleared */
 	ret = hda_dsp_wait_d0i3c_done(sdev);
 	if (ret < 0) {
 		dev_err(bus->dev, "CIP timeout before D0I3C update!\n");
-		return ret;
+		goto exit;
 	}
 
 	/* Update D0I3C register */
-	value = d0_substate == SOF_DSP_D0I3 ? SOF_HDA_VS_D0I3C_I3 : 0;
-	snd_hdac_chip_updateb(bus, VS_D0I3C, SOF_HDA_VS_D0I3C_I3, value);
+	snd_hdac_chip_updateb(bus, VS_D0I3C, SOF_HDA_VS_D0I3C_I3, 0);
 
 	/* Wait for cmd in progress to be cleared before exiting the function */
 	ret = hda_dsp_wait_d0i3c_done(sdev);
 	if (ret < 0) {
 		dev_err(bus->dev, "CIP timeout after D0I3C update!\n");
-		return ret;
+		goto exit;
 	}
 
 	dev_vdbg(bus->dev, "D0I3C updated, register = 0x%x\n",
 		 snd_hdac_chip_readb(bus, VS_D0I3C));
 
-	if (d0_substate == SOF_DSP_D0I0)
-		flags = HDA_PM_PPG;/* prevent power gating in D0 */
-	else
-		flags = HDA_PM_NO_DMA_TRACE;/* disable DMA trace in D0I3*/
+	flags = HDA_PM_PPG;/* prevent power gating in D0 */
 
 	/* sending pm_gate IPC */
 	ret = hda_dsp_send_pm_gate_ipc(sdev, flags);
-	if (ret < 0)
+	if (ret < 0) {
 		dev_err(sdev->dev,
 			"error: PM_GATE ipc error %d\n", ret);
+		goto exit;
+	}
 
+	/* success, update power state */
+	hda->dsp_power_state = SOF_DSP_STATE_D0;
+
+exit:
+	mutex_unlock(&hda->ps_mutex);
 	return ret;
 }
 
+int hda_dsp_set_d0i3(struct snd_sof_dev *sdev, u32 flags)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	int ret = 0;
+
+	mutex_lock(&hda->ps_mutex);
+
+	/* return if in D3 */
+	if (hda->dsp_power_state == SOF_DSP_STATE_D3) {
+		dev_vdbg(bus->dev, "%s dsp power state already in %d\n",
+			 __func__, hda->dsp_power_state);
+		goto exit;
+	}
+
+	/* sending pm_gate IPC */
+	ret = hda_dsp_send_pm_gate_ipc(sdev, flags);
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"error: PM_GATE ipc error %d\n", ret);
+		goto exit;
+	}
+
+	/* return if already in D0I3 */
+	if (hda->dsp_power_state == SOF_DSP_STATE_D0I3)
+		goto exit;
+
+	/* Write to D0I3C after Command-In-Progress bit is cleared */
+	ret = hda_dsp_wait_d0i3c_done(sdev);
+	if (ret < 0) {
+		dev_err(bus->dev, "CIP timeout before D0I3C update!\n");
+		goto exit;
+	}
+
+	/* Update D0I3C register */
+	snd_hdac_chip_updateb(bus, VS_D0I3C, SOF_HDA_VS_D0I3C_I3,
+			      SOF_HDA_VS_D0I3C_I3);
+
+	/* Wait for cmd in progress to be cleared before exiting the function */
+	ret = hda_dsp_wait_d0i3c_done(sdev);
+	if (ret < 0) {
+		dev_err(bus->dev, "CIP timeout after D0I3C update!\n");
+		goto exit;
+	}
+
+	dev_vdbg(bus->dev, "D0I3C updated, register = 0x%x\n",
+		 snd_hdac_chip_readb(bus, VS_D0I3C));
+
+	hda->dsp_power_state = SOF_DSP_STATE_D0I3;
+
+exit:
+	mutex_unlock(&hda->ps_mutex);
+	return ret;
+}
```

Comment:
@keyonjie I'm going to leave a long comment about the changes in this commit.

1. The checks for power state inside the set_D0i0/set_D0i3 are a bit of an overkill. In reality, there are only 2 options: enter D0i3 or exit D0i3. Entering is always from D0, you'd never really see the case where you're trying to enter D0i3 from D3. Likewise, exiting D0i3 is only possible if you're already in D0i3. Otherwise, it would be a system resume or a runtime resume. 

2. I still do not appreciate the mutex. The commands to enter/exit D0i3 are already protected by waiting for the CIP timeout. Do we really need a mutex for just setting the hda->dsp_power_state? This is not really a register write. Maybe @kv2019i and @plbossart can correct me if I am wrong

3. The bit masks for the power states are also not pretty. There are 3 states, why cant we just simply use an enum instead of setting a D0 bit along with I3 bit? Makes it a lot easier to understand

Reply:
> @keyonjie I'm going to leave a long comment about the changes in this commit.
> 
> 1. The checks for power state inside the set_D0i0/set_D0i3 are a bit of an overkill. In reality, there are only 2 options: enter D0i3 or exit D0i3. Entering is always from D0, you'd never really see the case where you're trying to enter D0i3 from D3. Likewise, exiting D0i3 is only possible if you're already in D0i3. Otherwise, it would be a system resume or a runtime resume.

That's true, but since the enum dsp_power_state can have the D3 value, we only implement the helpers here, we can't make sure the caller will use it correctly always?

Safe check is a compromise, some scenarios might never happen(e.g. D3->D0I3 entry, D3->D0I3 exit), some might happen via error used(e.g. D0I3->D0I3 entry, D0->D0I3 exit).

I preferred to add these check here to avoid critical errors happen or disordering the state(e.g. if we do try doing D3->D0I3), but I am fine if both you and @plbossart think we should remove the D3 check.

> 2. I still do not appreciate the mutex. The commands to enter/exit D0i3 are already protected by waiting for the CIP timeout. Do we really need a mutex for just setting the hda->dsp_power_state? This is not really a register write. Maybe @kv2019i and @plbossart can correct me if I am wrong

I introduced the mutex as I think we could have race in setting and getting this dsp_power_state, especially considering that the hda_dsp_set_d0i3() could be a delayed work which could run in another thread.

> 3. The bit masks for the power states are also not pretty. There are 3 states, why cant we just simply use an enum instead of setting a D0 bit along with I3 bit? Makes it a lot easier to understand

The key point here is that sometimes the caller(e.g. after a normal IPC sent) actually don't know which state it should put the DSP to, I introduce a command SOF_DSP_STATE_SET_AUTO here for these scenarios. So the rule here is only using command when calling hda_dsp_set_power_state(), and leave the DSP state change decision to the dsp state management logic(inside hda-dsp.c) only.

```
enum sof_dsp_state {
	SOF_DSP_STATE_D3 = 0,
	SOF_DSP_STATE_D0 = SOF_DSP_STATE_BIT_D0,
	SOF_DSP_STATE_D0I3 = SOF_DSP_STATE_BIT_D0 | SOF_DSP_STATE_BIT_I3,
};

enum sof_dsp_state_cmd {
	SOF_DSP_STATE_SET_D3 = SOF_DSP_STATE_D3,
	SOF_DSP_STATE_SET_D0 = SOF_DSP_STATE_D0,
	SOF_DSP_STATE_SET_D0I3 = SOF_DSP_STATE_D0I3,
	SOF_DSP_STATE_SET_AUTO = SOF_DSP_STATE_BIT_UPDATE,
};
```